### DISCUSSION TITLE
Wifi wrong?

### DISCUSSION THREAD

**Discussion by Rainer1969** [ohAnd/dtuGateway#35] [Q&A] [answered]
Hi all,
seems that I need some starter help.
My plan is to use the ESP to pull data from my brandnew Hoymiles HMS-800W-2T in a simple mode, but maybe I've  misunderstood something?
I think the ESP connects to the DTU and offers a simple webpage with the data (watts, etc.)
I'm working on Linux Mint and used this command to flash the ESP:
esptool -p /dev/ttyUSB0 write_flash 0x0000 dtuGateway_snapshot_1.6.0040_develop.bin
Since the OLED Display was mentinoed as to be optional, I have no display so far.

The ESP now opens a new WiFi AP dtuGateway_10197101 and I can conenct to it.
There is NO AP hoymilesGW_ like mentioned in documentation, step 1. first setup
Regarding documentation step (2), first setup, next step is to open http://192.168.4.1/ (or http://hoymilesgw.local/)
I can ping this address, but no website is opened - firefox reports network time limit error.
Firewall is switched off.
What I'm doing wrong?

**Comment by ohAnd** [comment]
Hi @Rainer1969 ,

sorry for the confusion - the recent snapshots have errors in this. (as you already noticed - the webserver is not coming up in first start mode).

Due to the last quick hacks a refactoring was needed and it is still in dev.

But you can give it a try as an early bird ;-) - especially when you only need the gateway function (reading dtu values and showing these on the webpage). With the refactoring there also comes multi-arch builds for ESP8266 and ESP32.

You can find it here -> [https://github.com/ohAnd/dtuGateway/tree/migrate_ESP32](https://github.com/ohAnd/dtuGateway/tree/migrate_ESP32) - regarding readme and so on.
dev firmware files are in the [actions section](https://github.com/ohAnd/dtuGateway/actions/workflows/feature_build.yml) located -> e.g. the last one as I'm writing this -> [https://github.com/ohAnd/dtuGateway/actions/runs/9738889777](https://github.com/ohAnd/dtuGateway/actions/runs/9738889777)

Please give a short feedback if it's working or also if not! ;-)

BR

**Comment by Rainer1969** [comment]
Hi Andreas / ohAnd,
thanks for your fast answer., I would be happy happy to be a "early bird", but I have a question:
I'm using ESP8266 - do you really mean I should flash this file from your link: "dtuGateway_snapshot_1.9.0030_migrate_ESP32" ?
If not where can I find the ESP8266 version?
Thank you + Nice evening!
BR;Rainer

**Reply by ohAnd** [reply]
:-) that is a zip file inside there are 2 bin files, one for ESP8266 and one for ESP32

**Comment by Rainer1969** [comment]
![Bildschirmfoto vom 2024-07-02 11-50-54](https://github.com/ohAnd/dtuGateway/assets/168187048/418d6588-423f-43de-bc33-586ce4e7ab7a)

Moin Andreas!
Successfully extracted and flashed - and webserver is running. Sorry to oversee the the .zip (but who thinks that in an file x is y ;-)
So far so good. Many thanks for opening my eyes.

What I don't get is, why I have to integrate the DTU in my network first. Currently my device is not yet in the cloud (and I don't really want to have it there)..

There is a bunch of reasons, why I would prefer contact the Wifi of the PUT directly:
* My DTU is located in my garden with almost no access to my Fritzbox
* I could place the ESP in the middle of distances, -> the ESP could send data to the internet, but not the DTU,
* I hate to be dependent on companys like Hoymiles, and here what I have to have?
* Smartphone with S-Miles App  with GPS Access AND access to my network AND  sim-card (I tried to do this with an old phone without SIM-card- and app was not working). 
* S-Miles account and resulting cloud integration of my device, and this is the worst.

 if anything is missing - nothing is working.
The reason to try ohAnd is to be INDEPENDENTLY from all the Hoymiles app and cloud stuff !?

Under the line I would like to raise the discussion, if it is not possible to contact directly the DTU. without having it integrated in the private network and S-Miles-Cloud.
Maybe there could be a solution where the ESP switches their Wifi each update cylce to the DTU - and the rest of time it's offering via webfrontend the data - via own Wifi as a server, or via existing network as a client, latter additionally with MTQQ and openhab support?
Nevertheless I try to do all the mandatory things above to connect the DTU with my network to be able to test further.
PS: I will be offline in the next days, so this will take time. 

Thank you again for providing  ohAnd and Best Regards,
Rainer

**Comment by ohAnd** [comment]
Moin Rainer!

I understand your idea / target setup...

As I started with the project, I connected the dtuGateway directly with the dtu over the dtu internal access point and read the values directly over serial port.
Therefore there is still a serial output implemented to do that.

As a suggestions, for a first setup on your side:
- dtu serve his own network "DTUBI-xxxxxxxxx"
- to this network you can connect with your phone (or whatever wifi client)
- find out which IP will be then used by your client (e.g. 10.10.100.xxx) - this should be show that the DTU will served at e.g. 10.10.100.1
- then you can connect with the dtuGateway (first start or already configured in your home network) and connect the dtuGateway directly with the "DTUBI-xxxxxxxxx" network with the needed password for the DTU AP
- then you can go back with your phone (wifi client) and connect to the DTU AP also again
- now DTU, dtuGateway and your phone should be all within the same local network of the DTU
- in dtuGateway you can now enter the server IP of the DTU in the dtu seetings (data for wireless accesspoint in this tab are currently not needed)
- after this the dtuGateway should connect to the DTU directly and show the current state (current wattage, etc.) via webpage
- => connection established ;-)
- next you could now connect dtuGateway with usb and grab the output data over serial port with a terminal
   - for those there is a [very small command interface](https://github.com/ohAnd/dtuGateway/blob/4424079d2f2efad6d1d73f4ea1da3b088532baff/src/dtuGateway.ino#L1048) (the debug surroundings are still there and has to be ignored)
      - dataFormatJSON 1 - giving a json formatted string with all data
      - dataFormatJSON 0 - giving a readable formatted output
 
so far... for this possibility.

The idea for changing the networks for a given time... in my opinion not suitable for the target application to use the powerSetting to get dynamic changes of the output power of ther inverter for e.g. surplus control. I want to have the control to change every second.

So, when it should go in this direction that you prefer, then I would suggest to connect two dtuGateways over serial (SPI, whatever) to use one as a connection to the DTU and one for the connection to your local network and both exchanging only the DTU relevant data.

btw. I don't know at this time, if there is any configuration or first setup of the DTU needed to be "in the internet" needed for the function, the only thing could be, that you have outdated firmware of the dtu or outdated grid config. But I would assume that without any setup (bring dtu ot internet/ hoymiles connection) it should be accessible as described above.

BR
Andreas

**Comment by Rainer1969** [comment]
Send long answer via reply  email - I hope it will appear here.

**Comment by Frankyistda** [comment]
Hi together!

Sorry, but no matter whether I use dtuGateway_ESP32_snapshot_1.9.0030_migrate_ESP32.bin or dtuGateway_ESP32_snapshot_1.9.0048_develop.bin, no new access point is displayed.

If I work with the old but useless opendtu-generic.factory.bin, I immediately have an OpenDTU-xxx access point.

As hardware I have an ESP32-D0WD-V3 board, a radio module E01-ML01DP5 V3.2 based on nRF24L01+ and the 1.3” display SH1106.

Who can help? @ohAnd Andreas? Hope so!

Greetings
Frank

**Reply by ohAnd** [reply]
Hi @Frankyistda ,

I've seen no issues in this direction for the ESP32 binarys.
Unfortunately I have no possibility to check in detail in the next few days.

Maybe you can check meanwhile what the serial com is telling, directly after reset/ start... (with the latest snapshot 1.9.0048_develop)
There should be the info of the flashed firmware and also a detailed output about the current user settings (these should be after first flashing the defaults) - especially whether the controller is in initMode/ APStartMode.
And if that the case, also the info that the "dtuGateway_xxx" AP will be started.

Maybe with this info we can find the problem.

btw.
dtuGateway is not a replacement for opendtu - thus there is no direct comparison given.
Your hardware ESP32 and the display with SH1106 should be usually no problem - the radio module will not be needed.
 
So, please give the serial com a chance.

BR
Andreas

**Reply by ohAnd** [reply]
Hi @Frankyistda ,

I've checked that with a fresh ESP32 WROOM S2 - the access point will be provided and after booking in, I can open the config page at 192.168.1.4 to configure my local wifi...

The serial output should be looking like this below.
Can you please check on your side the serial output?

Thx

```
10:59:03.852 > Booting - with firmware version 1.9.0048_develop
10:59:03.856 > ------------------------------------------------------------------
10:59:03.874 > UserConfigManager::loadConfig - config loaded from json: /userconfig.json
10:59:03.881 > UserConfigManager::begin - Config loaded successfully
10:59:03.895 > UserConfigManager::loadConfig - config loaded from json: /userconfig.json
10:59:03.901 >
10:59:03.901 > --------------------------------------
10:59:03.905 > Configuration loaded from config file: '/userconfig.json'
10:59:03.910 > init (wifiAPstart):      1
10:59:03.912 > wifi ssid:               mySSID
10:59:03.914 > wifi pass:               myPassword
10:59:03.916 >
10:59:03.916 > dtu update time:         31
10:59:03.918 > dtu host:                192.168.0.254
10:59:03.920 > dtu ssid:                DTUBI-12345678
10:59:03.922 > dtu pass:                dtubiPassword
10:59:03.924 > cloud pause active:      1
10:59:03.926 > cloud pause time:        40
10:59:03.929 >
10:59:03.929 > openhab host:            192.168.1.100
10:59:03.931 > openhab item prefix:     inverter
10:59:03.934 > openhab binding active:  0
10:59:03.936 >
10:59:03.937 > mqtt host:               192.168.1.100
10:59:03.940 > mqtt port:               1883
10:59:03.941 > mqtt TLS:                0
10:59:03.942 > mqtt user:               dtuuser
10:59:03.944 > mqtt pass:               dtupass
10:59:03.946 > mqtt topic:              dtu_309xxxx
10:59:03.949 > mqtt binding active:     0
10:59:03.950 > mqtt HA autoDiscovery:   0
10:59:03.952 >
10:59:03.953 > remoteDisplay:           0
10:59:03.954 > update channel:          0
10:59:03.956 > display connected:       1
10:59:03.959 > --------------------------------------
10:59:04.287 > TFT display initialized
10:59:04.289 >
10:59:04.289 > +++ device in 'first start' mode - have to be initialized over own served wifi +++
10:59:04.297 >
10:59:10.401 > WIFI_SCAN:        done: 7 wifi's found
10:59:10.415 >
10:59:10.416 >  +++ serving access point with SSID: 'dtuGateway_309xxxx' +++
10:59:10.421 >
10:59:10.421 > AP IP address: 192.168.4.1
10:59:10.436 > Ready! Open http://dtuGateway.local in your browser
10:59:10.440 > TFT display - showing factory mode
10:59:10.473 > config saved to json: /userconfig.json
10:59:10.482 > WEB:              setup webserver
10:59:10.485 >
```

**Comment by Frankyistda** [comment]
Sorry Andreas for my late reply, but apparently my message from a few days ago didn't make it here. I must have done something wrong or it has ended up somewhere wrong!

Incidentally, my board has **esp32doit-devkit-v1** printed on it and the module on it is an **ESP-WROOM-32**. The old, now useless openDTU displayed it as ESP32-D0WD-V3, but that must not be correct.

Now to my problem: I have only ever flashed the finished BINs with the Flash Tool.
I therefore have no idea how to get the serial com output. The board is connected via USB.

Do I probably need Putty? Maybe you can help me out...

Thank you very much!

Greetings
Frank

PS: Is English mandatory or can I also answer in German. Then I can save myself the DeepL translation. :-)

**Reply by ohAnd** [reply]
Hi @Frankyistda ,

klar geht auch ;-) ...
Jep, Putty ist ein einfacher Weg - ESP mit USB an PC (am Beispiel Windows) -> damit sollte automatisch der COM-Treiber (z.B. CP210x oder CH340) installiert werden.
Über "Geräte und Drucker" den COM-Port ausfindig machen. In Putty "Serial" wählen und den COM-Port z.b. "COM2" eintragen. Dann starten und die Debug-Infos sollten sichtbar werden.
Über den Reset/ enable-Button kannst Du dann auch einen Reboot auslösen um zur oben gezeigten Output zu kommen...

Gruss

**Comment by Frankyistda** [comment]
Hallo Andrea @ohAnd ,

vielen Dank!

Ich habe das Board jetzt noch mal neu geflashed mit Deiner dtuGateway_ESP32_snapshot_1.9.0049_develop.bin Datei:

![flash](https://github.com/user-attachments/assets/07f81c7f-52bf-41fb-93a2-abf3c15acfdd)

Die Ausgabe ist:

D:\users\Frank\_Franks Dateien\Technische Downloads\_Haus\Photovoltaik\OpenDTU_AhoyDTU\Tool für WiFi Wechselrichter\dtuGateway_ESP32_snapshot_1.9.0049_develop.bin
test offset :  0 0x0
case ok
.
Uploading stub...
Running stub...
Stub running...
Changing baud rate to 115200
Changed.
NO XMC flash  detected!
FLASH_CRYPT_CNT 0
ABS_DONE_0 False
Compressed 1463552 bytes to 866381...

 is stub and send flash finish

-----------------------------------------------------------------------------------------------------------------------------------

Wenn ich jetzt mit Putty an den seriell COM gehe, dann läuft dort nur immer durch:

ets Jul 29 2019 12:21:46

rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
invalid header: 0x2068746f
invalid header: 0x2068746f
invalid header: 0x2068746f
invalid header: 0x2068746f
invalid header: 0x2068746f
invalid header: 0x2068746f
invalid header: 0x2068746f
invalid header: 0x2068746f
ets Jul 29 2019 12:21:46

rst:0x10 (RTCWDT_RTC_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
invalid header: 0x2068746f
invalid header: 0x2068746f
invalid header: 0x2068746f
invalid header: 0x2068746f
invalid header: 0x2068746f
invalid header: 0x2068746f
invalid header: 0x2068746f
invalid header: 0x2068746f
ets Jul 29 2019 12:21:46

rst:0x10 (RTCWDT_RTC_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
invalid header: 0x2068746f
invalid header: 0x2068746f
invalid header: 0x2068746f
invalid header: 0x2068746f
invalid header: 0x2068746f
invalid header: 0x2068746f
invalid header: 0x2068746f
invalid header: 0x2068746f
ets Jul 29 2019 12:21:46

**Reply by ohAnd** [reply, marked as answer]
Hallo Frank ( @Frankyistda ),

... now I got it ... ;-)
Den Ablauf für das erste ESP32 flashen hatte ich noch nicht in der readme beschrieben... Da ich bei mir direkt mit platformio (via esptool) flashe, kam das Problem auch noch nicht auf ;-)

Daher nun hier der Ablauf für das erstmalige flashen eines ESP32 mit dem "flash download tool".
Es werden 4 Dateien benötigt:
- [dtu_gateway_flash_basics.zip](https://github.com/user-attachments/files/16836687/dtu_gateway_flash_basics.zip)
  - bootloader
  - boot app
  - partitions
- firmware (für das spätere update wird für das webupdate nur diese benötigt)

Die ersten 3 files werde ich dann zukünftig im repo verlinken.

flash download tool:
![image](https://github.com/user-attachments/assets/179c7a33-5fd2-476f-8213-1d5ab4db0171)

sollte dann so aussehen:
```
Uploading stub...
Running stub...
Stub running...
Changing baud rate to 115200
Changed.
FLASH_CRYPT_CNT 0
ABS_DONE_0 False
test offset :  4096 0x1000
case ok
test offset :  32768 0x8000
case ok
test offset :  57344 0xe000
case ok
test offset :  65536 0x10000
case ok
.
Uploading stub...
Running stub...
Stub running...
Changing baud rate to 921600
Changed.
NO XMC flash  detected!
FLASH_CRYPT_CNT 0
ABS_DONE_0 False
Compressed 17536 bytes to 12202...
Compressed 3072 bytes to 146...
Compressed 8192 bytes to 47...
Compressed 1462736 bytes to 865940...

 is stub and send flash finish
```

Anschliessend Enable drücken oder USB ab- und wieder anstecken, dann sollte im COM folgendes zu sehen sein:

```
13:12:16.071 > Booting - with firmware version 1.9.0048_develop
13:12:16.074 > ------------------------------------------------------------------
13:12:16.080 > [    22][E][vfs_api.cpp:105] open(): /littlefs/userconfig.json does not exist, no permits for creation
13:12:16.090 > UserConfigManager::begin - First run: Initializing config
13:12:16.103 > config saved to json: /userconfig.json
13:12:16.131 > UserConfigManager::loadConfig - config loaded from json: /userconfig.json
13:12:16.138 >
13:12:16.138 > --------------------------------------
13:12:16.141 > Configuration loaded from config file: '/userconfig.json'
13:12:16.146 > init (wifiAPstart):      1
13:12:16.148 > wifi ssid:               mySSID
13:12:16.150 > wifi pass:               myPassword
13:12:16.152 >
13:12:16.152 > dtu update time:         31
13:12:16.155 > dtu host:                192.168.0.254
13:12:16.156 > dtu ssid:                DTUBI-12345678
13:12:16.159 > dtu pass:                dtubiPassword
13:12:16.161 > cloud pause active:      1
13:12:16.163 > cloud pause time:        40
13:12:16.165 >
13:12:16.165 > openhab host:            192.168.1.100
13:12:16.168 > openhab item prefix:     inverter
13:12:16.171 > openhab binding active:  0
13:12:16.173 >
13:12:16.173 > mqtt host:               192.168.1.100
13:12:16.177 > mqtt port:               1883
13:12:16.177 > mqtt TLS:                0
13:12:16.178 > mqtt user:               dtuuser
13:12:16.180 > mqtt pass:               dtupass
13:12:16.183 > mqtt topic:              dtu_12345678
13:12:16.185 > mqtt binding active:     0
13:12:16.187 > mqtt HA autoDiscovery:   0
13:12:16.190 >
13:12:16.190 > remoteDisplay:           0
13:12:16.192 > update channel:          0
13:12:16.194 > display connected:       0
13:12:16.196 > --------------------------------------
13:12:16.505 > OLED display:     initialized
13:12:16.507 >
13:12:16.509 > +++ device in 'first start' mode - have to be initialized over own served wifi +++
13:12:16.515 >
13:12:22.620 > WIFI_SCAN:        done: 9 wifi's found
13:12:22.645 > 
13:12:22.646 >  +++ serving access point with SSID: 'dtuGateway_309xxxx' +++
13:12:22.651 >
13:12:22.651 > AP IP address: 192.168.4.1
13:12:22.661 > Ready! Open http://dtuGateway.local in your browser
13:12:22.666 > OLED display:     showing factory mode
13:12:22.692 > config saved to json: /userconfig.json
13:12:22.698 > WEB:              setup webserver
```

Hoffe das es in deiner Umgebung so nachvollziehbar ist.

Gruss
Andreas

**Comment by checkersky** [comment]
Hi, was auch gehen sollte:

Für den ersten Flash reicht es auch eine Aktuelle bin ( also z.b die aktuelle Version oder ein Snapshot ) zu nehmen und die mit dem "flash download tool" ab adresse 0x0 zuflashen.
Wichtig ist da  "DoNotChgBin" an ist.

So spart man sich das mit  bootloader, boot app , partitions usw.

**Reply by ohAnd** [reply]
Hi,

danke für den Hinweis.
Hab mich nicht mit allen Details beschäftigt ... 

Aber für das ESP32 Projekt habe ich die partions angepasst ("min_spiffs.csv" -> [see](https://github.com/ohAnd/dtuGateway/blob/fbf144517225fd2873541bce2e6c03f3f1a10ca9/platformio.ini#L36)), damit das FS ein wenig kleiner ist und damit mehr Platz für die Applikation bleibt. Daher wird mindestens die partions.bin relevant sein - denke ich.

Daher ist es in Abhängigkeit dessen, was vorher auf dem Controller geflashed wurde bzw. welche default config enthalten ist, mglw. sinnvoll diesen einmaligen "langen" Weg zu nutzen - um mgl. Folgefehler auszuschliessen.


btw. was mir hier unklar ist, warum mit der Zieladresse 0x0 die Firmware trotzdem an die richtige Stelle in den Flash geschrieben wird...

Gruss

**Comment by Frankyistda** [comment]
Hallo Andreas @ohAnd ,

heureka! Es hat geklappt! Vielen Dank für Deine Hilfe! 

Worüber ich noch gestolpert bin: Ich habe nach dem Verbinden des WLANs des ESP32 Moduls mit einem Windows 11 Laptop (mein Android Smartphone wollte keine Verbindung herstellen) und Aufruf von: http://192.168.4.1 mein WiFi Netz und passendes Passwort eingetragen.

Dann bin ich zum 2. Reiter DTU gewechselt und habe dort die IP-Adresse des Mikrowechselrichters eingetragen, aber in der Seriellen Konsole von Putty kam trotzdem weiterhin: "DTUinterface:    getDataUpdate - ERROR - not connected to DTU!"

Mein Fehler: Die Verbindung des ESP32 hatte sich in der Zwischenzeit bereits mit meinem Haus WiFi verbunden und die Änderung an dem Laptop nicht mehr mitbekommen. Nach einem Reboot des ESPs konnte ich das dann auch ablesen.

Die Änderung über die Web GUI hat dann anstandslos geklappt und nun werden die Daten in der GUI und im Display angezeigt!

Tolle Sache! Bin total happy!

Genommen habe ich Version 1.9.0049, habe aber gesehen, dass Du bereits eine v1.9.0052 eingestellt hast. Die betrifft aber nur das OLED Display und nicht mein SH1106 Display, richtig?

Eine Anregung für Verbesserung hätte ich noch: Bei dem bisherigen openDTU Programm, konnte man die Helligkeit der Anzeige dimmen und wenn der Wechselrichter abends offline ging, auch das Display ausgehen lassen. Und man konnte einen Modus gegen "Einbrennen" aktivieren, so dass die Schrift sich immer ein wenig bewegte.

Wenn Du mal Langeweile hast und das auch als sinnvoll erachtest, würde ich mich darüber freuen. (ja, so ist das, erst reichst Du den kleinen Finger und dann nehme ich gleich die ganze Hand, sorry ;-)

Aber ganz im Ernst, tolle Arbeit hast Du da vollbracht! 1A mit Mappe!

Gruß 
Frank

**Reply by ohAnd** [reply]
Freut mich, das es jetzt geklappt hat ... bzgl. Display dunkel siehe hier #45 ;-)

**Comment by Frankyistda** [comment]
Was bedeutet in der Anzeige eigentlich ab und zu das Smiley? Das er Daten nicht abrufen kann, wegen der Cloud Pause?

![20240907_091523](https://github.com/user-attachments/assets/61d00a8d-9d83-46d6-9b6b-40f43ae490e4)

**Reply by ohAnd** [reply]
Hi @Frankyistda ,

Jep, das zeigt die aktive "Cloud Pause" an. Hier wird für die eingestellte Zeit (advanced config) nicht von der dtu abgerufen, damit der Upload von der dtu zur hoymiles cloud nicht gestört wird.

Wenn nicht benötigt, kann es über die settings abgeschaltet werden...

Gruss

**Comment by checkersky** [comment]
Hi, das zeigt dir an, dass er gerade ein Cloud Update macht

**Comment by Frankyistda** [comment]
Hallo Andreas @ohAnd,

meinst Du, Du kannst in dem nächsten Release vielleicht die Ausgabe für den Totalertrag rechts unten im Display des OLEDs um eine Nachkommastelle ergänzen? 

Ist anscheinend in der display.cpp die Zeile mit: 
u8g2.drawStr(3 + 18 * 4 + offset_x, 56 + offset_y, ("t: " + String(lastDisplayData.totalYieldTotal, 0) + " kWh").c_str());

Das wäre schön! Ich habe doch nur so eine kleine PV-Anlage und es wird Herbst... 

Gruß
Frank
